crates/core/src/client/messages.rs

```diff
@@ -28,46 +30,67 @@ pub trait ToProtocol {
     fn to_protocol(self, protocol: Protocol) -> Self::Encoded;
 }
 
+fn serialize_gzip(bytes: &[u8]) -> Vec<u8> {
+    let mut encoder = GzEncoder::new(Vec::new(), flate2::Compression::fast());
+    encoder.write_all(bytes).unwrap();
+    encoder
+        .finish()
+        .expect("Failed to Gz compress `SubscriptionUpdateMessage`")
+}
+
+fn serialize_brotli(bytes: &[u8]) -> Vec<u8> {
+    let reader = &mut &bytes[..];
+
+    // TODO(perf): Compression should depend on message size and type.
+    //
+    // SubscriptionUpdate messages will typically be quite large,
+    // while TransactionUpdate messages will typically be quite small.
+    //
+    // If we are optimizing for SubscriptionUpdates,
+    // we want a large buffer.
+    // But if we are optimizing for TransactionUpdates,
+    // we probably want to skip compression altogether.
+    //
+    // For now we choose a reasonable middle ground,
+    // which is to compress everything using a 32KB buffer.
+    const BUFFER_SIZE: usize = 32 * 1024;
+    // Again we are optimizing for compression speed,
+    // so we choose the lowest (fastest) level of compression.
+    // Experiments on internal workloads have shown compression ratios between 7:1 and 10:1
+    // for large `SubscriptionUpdate` messages at this level.
+    const COMPRESSION_LEVEL: u32 = 1;
+    // The default value for an internal compression parameter.
+    // See `BrotliEncoderParams` for more details.
+    const LG_WIN: u32 = 22;
+
+    let mut encoder = CompressorReader::new(reader, BUFFER_SIZE, COMPRESSION_LEVEL, LG_WIN);
+
+    let mut out = Vec::new();
+    encoder
+        .read_to_end(&mut out)
+        .expect("Failed to Brotli compress `SubscriptionUpdateMessage`");
+    out
+}
+
```

Comment:
The combination of both looks like will benefit from a shared trait or similar so we encode the invariants described in the comments and the `const` used there.


Like:

```rust
struct Compresor {
    algo: Compression,
}

impl Compresor {
    const BUFFER_SIZE: usize = 32 * 1024;
    // Separated values by algorithm
    const BROTOLI_LG_WIN: u32 = 22;
    
    fn compress(&self, data: &[u8]) -> Vec<u8> {
        match self.algo {
            Compression::None => data.to_vec(),
            Compression::Gzip => {
               ...
            }
            Compression::Brotli => {
             ...
            }
        }
    }
}
```

Reply:
It also could be moved all the compression into another crate or module